File: C++/MergekSortedLists.cc
// Merge k sorted linked lists and return it as one sorted list.
// Analyze and describe its complexity.

class Solution {
    class Compare {
    public:
        bool operator() (const ListNode* lhs, const ListNode* rhs) {
          return lhs->val > rhs->val;
        }
    };
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode head(0);
        ListNode* cur = &head;
        priority_queue<ListNode*, vector<ListNode*>, Compare> min_heap;
        for (auto l : lists) {
            if (l) min_heap.push(l);
        }
        while (!min_heap.empty()) {
            ListNode* l = min_heap.top();
            cur->next = l;
            min_heap.pop();
            if (l->next) {
                min_heap.push(l->next);
            }
            cur = cur->next;
        }
        return head.next;
    }
};

File: Java/MergeKSortedLists.java
// Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.

public class MergeKSortedLists {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        }
        PriorityQueue<ListNode> pq = new PriorityQueue<>(lists.length, new ListNodeComparator());
        for (ListNode l : lists) {
            if (l != null) {
                pq.add(l);
            }
        }
        ListNode head = new ListNode(0);
        ListNode cur = head;
        while (!pq.isEmpty()) {
            ListNode l = pq.poll();
            cur.next = l;
            cur = cur.next;
            if (l.next != null) {
                pq.add(l.next);
            }
        }
        return head.next;
    }

    private class ListNodeComparator implements Comparator<ListNode> {
        @Override
        public int compare(ListNode l1, ListNode l2) {
            return Integer.compare(l1.val, l2.val);
        }
    }
}
